### PR TITLE
fix(wallet): disable solana dapps in android prod

### PR DESCRIPTION
**Pull request by Pavneet-Sing**
Resolves https://github.com/brave/brave-browser/issues/26804

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Solana dapps should work on all Android build channels except on release

[sol-release.webm](https://user-images.githubusercontent.com/11755381/202293890-b11a1f7f-dbe3-48da-a82b-1abe76349723.webm)

### CODE REVIEW

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc [hunk 1]
```diff
@@ -11,6 +11,8 @@
 #include "base/feature_list.h"
 #include "brave/components/brave_wallet/common/features.h"
 #include "brave/components/brave_wallet/renderer/v8_helper.h"
+#include "chrome/common/channel_info.h"
+#include "components/version_info/version_info.h"
```
Comment (SergeyZhukovsky) [inline renderer/brave_wallet/brave_wallet_render_frame_observer.cc:15]: deps should be added to https://github.com/brave/brave-core/blob/master/renderer/brave_wallet/BUILD.gn#L17
for `components/version_info`

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc [hunk 2]
```diff
@@ -100,6 +102,11 @@ void BraveWalletRenderFrameObserver::DidClearWindowObject() {
           brave_wallet::features::kBraveWalletSolanaProviderFeature) &&
       web_frame->GetDocument().IsDOMFeaturePolicyEnabled(context, "solana") &&
       dynamic_params.brave_use_native_solana_wallet) {
+    // Disable Solana dapps support on android release channel
+    if (BUILDFLAG(IS_ANDROID) &&
```
Comment (SergeyZhukovsky) [inline renderer/brave_wallet/brave_wallet_render_frame_observer.cc:106]: that's a mess, `BUILDFLAG(IS_ANDROID)` is a preprocessor check, the build will fail.

Comment (SergeyZhukovsky) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:106]:
```
#if BUILDFLAG(IS_ANDROID)
  if (chrome::GetChannel() == version_info::Channel::STABLE)
      return;
#endif
```

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc [hunk 3]
```diff
@@ -11,6 +11,8 @@
 #include "base/feature_list.h"
 #include "brave/components/brave_wallet/common/features.h"
 #include "brave/components/brave_wallet/renderer/v8_helper.h"
```
Comment (SergeyZhukovsky) [inline renderer/brave_wallet/brave_wallet_render_frame_observer.cc:13]: #include "build/buildflag.h"

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 1 above)
Comment (Pavneet-Sing) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:16]: added

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 2 above)
Comment (Pavneet-Sing) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc]: The local build worked though was thinking of doing the same, updated

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 3 above)
Comment (Pavneet-Sing) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:13]: updated, as above, why the system doesn't pin point if a required header file should be added/missing?

**Comment by Pavneet-Sing** [comment]
I am getting 
```

[2022-11-17T10:22:28.544Z] ERROR in /Users/jenkins/jenkins/workspace/pr-brave-browser-disable-solana-dapps-in-release-ios/src/brave/renderer/brave_wallet/brave_wallet_render_frame_observer.cc

[2022-11-17T10:22:28.544Z]   Illegal include: "components/version_info/channel.h"

[2022-11-17T10:22:28.544Z]     Because of no rule applying
```
thought https://github.com/brave/brave-core/blob/9fb60b8e2d6a98dde1d1df51bb7ba0fee217cd2d/renderer/brave_wallet/DEPS#L3 should fix it.

Can anyone with native c++ expertise explain or fix this issue?

**Comment by SergeyZhukovsky** [comment]
> I am getting
> 
> ```
> 
> [2022-11-17T10:22:28.544Z] ERROR in /Users/jenkins/jenkins/workspace/pr-brave-browser-disable-solana-dapps-in-release-ios/src/brave/renderer/brave_wallet/brave_wallet_render_frame_observer.cc
> 
> [2022-11-17T10:22:28.544Z]   Illegal include: "components/version_info/channel.h"
> 
> [2022-11-17T10:22:28.544Z]     Because of no rule applying
> ```
> 
> thought
> 
> https://github.com/brave/brave-core/blob/9fb60b8e2d6a98dde1d1df51bb7ba0fee217cd2d/renderer/brave_wallet/DEPS#L3
> 
> should fix it (not sure if there's an iOS specific file should be added).
> Can anyone with native c++ and chromium expertise explain or fix this issue?

I will take care of that PR. Don't worry about it, enjoy your days off!

File: renderer/brave_wallet/DEPS [hunk 4]
```diff
@@ -1,5 +1,6 @@
 include_rules = [
   "+brave/common",
+  "+brave/components/version_info",
```
Comment (goodov) [inline renderer/brave_wallet/DEPS:3]: this is not needed.

Comment (SergeyZhukovsky) [reply renderer/brave_wallet/DEPS:3]: yup, going to replace it to `"+components/version_info",`

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc [hunk 5]
```diff
@@ -11,6 +11,9 @@
...
+#include "build/buildflag.h"
+#include "chrome/common/channel_info.h"
+#include "components/version_info/channel.h"
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:16]: @goodov as per @Pavneet-Sing 's question do you know do we have something that is under development to detect `include` cases? (a missed include, unnecessary include and etc).

[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:16]:
```
../../brave/renderer/brave_wallet/brave_wallet_render_frame_observer.cc:107:47: error: incomplete type 'version_info::Channel' named in nested name specifier
    if (chrome::GetChannel() == version_info::Channel::STABLE)
```

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 1 above)
Comment (SergeyZhukovsky) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc]: `npm run gn_check` should fail if there are deps problems.

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 3 above)
Comment (SergeyZhukovsky) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:13]: Asked it here https://github.com/brave/brave-core/pull/15979#discussion_r1025639750

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (goodov) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:16]: > ```
> ../../brave/renderer/brave_wallet/brave_wallet_render_frame_observer.cc:107:47: error: incomplete type 'version_info::Channel' named in nested name specifier
>     if (chrome::GetChannel() == version_info::Channel::STABLE)
> ```

my bad. it's forward declared.

> @goodov as per @Pavneet-Sing 's question do you know do we have something that is under development to detect `include` cases? (a missed include, unnecessary include and etc).

nope, but through upstream presubmits we're about to launch, we will be able to cover some of includes that upstream has checks for.
We can also add our own checks to cover most frequent one that are usually lost.

**Review by goodov — APPROVED** [review]

**Comment by srirambv** [comment]
Verification passed on 1.47.85 x64 Nightly build
- Verified Solana Dapp support is enabled by default on Nightly build

https://user-images.githubusercontent.com/17010094/202738732-4682f93c-7532-46f2-b8b2-5559bbcd12a7.mp4

**Pull request merged**